File: packages/core/src/analysis/params.ts
/**
 * Parameter counting.
 *
 * `total` is every trainable weight. `active` is what a single token actually
 * uses, which differs from `total` only for mixture-of-experts designs.
 * `nonEmbedding` is the number that the 2N/6N FLOPs rules are written against.
 */

import type { Doc, SymbolTable } from "../ir/types.js";
import { flatten, type FlatNode, type FlatResult } from "./flatten.js";

export interface ParamsResult {
  /** Every trainable weight. */
  total: number;
  /** Weights a single token actually uses. Differs from total only for MoE. */
  active: number;
  embedding: number;
  head: number;
  nonEmbedding: number;
  /** Active parameters excluding the embedding table: the N in the 2N rule. */
  nonEmbeddingActive: number;
  byPath: Record<string, number>;
  byCategory: Record<string, number>;
  byType: Record<string, number>;
  errors: string[];
}

const EMBEDDING_TYPES = new Set(["embedding", "pos_embedding"]);

export function countParams(doc: Doc, symbols: SymbolTable, flat?: FlatResult): ParamsResult {
  const f = flat ?? flatten(doc, symbols);
  const res: ParamsResult = {
    total: 0,
    active: 0,
    embedding: 0,
    head: 0,
    nonEmbedding: 0,
    nonEmbeddingActive: 0,
    byPath: {},
    byCategory: {},
    byType: {},
    errors: [...f.errors],
  };

  for (const node of f.nodes) {
    const per = paramsOfNode(node, res.errors);
    if (per === null) continue;
    const total = per * node.multiplier;
    const active = per * node.activeMultiplier;
    if (total === 0) continue;

    res.total += total;
    res.active += active;
    res.byPath[node.path] = total;
    res.byCategory[node.category] = (res.byCategory[node.category] ?? 0) + total;
    res.byType[node.type] = (res.byType[node.type] ?? 0) + total;

    if (EMBEDDING_TYPES.has(node.type)) res.embedding += total;
    else if (node.type === "lm_head") res.head += total;
  }

  res.nonEmbedding = res.total - res.embedding;
  res.nonEmbeddingActive = res.active - res.embedding;
  return res;
}

function paramsOfNode(node: FlatNode, errors: string[]): number | null {
  if (!node.def.paramCount) return 0;
  try {
    const n = node.def.paramCount(node.resolved);
    if (!Number.isFinite(n)) {
      errors.push(`${node.path}: parameter count is not a finite number`);
      return null;
    }
    return n;
  } catch (e) {
    errors.push(`${node.path}: ${(e as Error).message}`);
    return null;
  }
}

/** Human-readable parameter count: 8.03B, 124.4M, 12.9K. */
export function formatCount(n: number): string {
  const abs = Math.abs(n);
  if (abs >= 1e12) return `${(n / 1e12).toFixed(2)}T`;
  if (abs >= 1e9) return `${(n / 1e9).toFixed(2)}B`;
  if (abs >= 1e6) return `${(n / 1e6).toFixed(1)}M`;
  if (abs >= 1e3) return `${(n / 1e3).toFixed(1)}K`;
  return String(n);
}

File: packages/core-go/analysis/params.go
package analysis

import (
	"math"
)

// ParamsResult is what the design weighs.
//
// Total is every trainable weight; Active is what a single token actually uses,
// which differs from Total only for a mixture of experts; NonEmbedding is the
// number the 2N and 6N FLOPs rules are written against.
type ParamsResult struct {
	Total        float64 `json:"total"`
	Active       float64 `json:"active"`
	Embedding    float64 `json:"embedding"`
	Head         float64 `json:"head"`
	NonEmbedding float64 `json:"nonEmbedding"`
	// NonEmbeddingActive is the active count without the embedding table: the N
	// in the 2N rule.
	NonEmbeddingActive float64 `json:"nonEmbeddingActive"`
	// Expert is the weights that live inside a moe_experts container. They are
	// most of a sparse model and they shard differently from the rest: expert
	// parallelism hands whole experts to whole devices, where tensor
	// parallelism splits every matrix across its group.
	Expert     float64            `json:"expert"`
	ByPath     map[string]float64 `json:"byPath"`
	ByCategory map[string]float64 `json:"byCategory"`
	ByType     map[string]float64 `json:"byType"`
	Errors     []string           `json:"errors"`
}

var embeddingTypes = map[string]bool{"embedding": true, "pos_embedding": true}

// CountParams adds up the weights, by path, category and type.
func CountParams(flat *FlatResult) *ParamsResult {
	res := &ParamsResult{
		ByPath:     map[string]float64{},
		ByCategory: map[string]float64{},
		ByType:     map[string]float64{},
		Errors:     append([]string{}, flat.Errors...),
	}

	for i := range flat.Nodes {
		node := &flat.Nodes[i]
		if node.Def.ParamCount == nil {
			continue
		}
		per := node.Def.ParamCount(node.Resolved)
		if !finite(per) {
			res.Errors = append(res.Errors, node.Path+": parameter count is not a finite number")
			continue
		}
		total := per * node.Multiplier
		active := per * node.ActiveMultiplier
		if total == 0 {
			continue
		}

		res.Total += total
		res.Active += active
		res.ByPath[node.Path] = total
		res.ByCategory[node.Category] += total
		res.ByType[node.Type] += total

		if embeddingTypes[node.Type] {
			res.Embedding += total
		} else if node.Type == "lm_head" {
			res.Head += total
		}
		if node.Expert {
			res.Expert += total
		}
	}

	res.NonEmbedding = res.Total - res.Embedding
	res.NonEmbeddingActive = res.Active - res.Embedding
	return res
}

// FormatCount is a parameter count as a person reads it: 8.03B, 124.4M, 12.9K.
func FormatCount(n float64) string {
	abs := math.Abs(n)
	switch {
	case abs >= 1e12:
		return JSToFixed(n/1e12, 2) + "T"
	case abs >= 1e9:
		return JSToFixed(n/1e9, 2) + "B"
	case abs >= 1e6:
		return JSToFixed(n/1e6, 1) + "M"
	case abs >= 1e3:
		return JSToFixed(n/1e3, 1) + "K"
	}
	return JSNumber(n)
}
